src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/filesystem/BitbucketSCMFile.java

```diff
@@ -0,0 +1,90 @@
+package com.atlassian.bitbucket.jenkins.internal.scm.filesystem;
+
+import com.atlassian.bitbucket.jenkins.internal.client.BitbucketFilePathClient;
+import com.atlassian.bitbucket.jenkins.internal.client.exception.NotFoundException;
+import jenkins.scm.api.SCMFile;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.StringUtils;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.util.Optional;
+
+import static jenkins.scm.api.SCMFile.Type.*;
+
+public class BitbucketSCMFile extends SCMFile {
+
+    private final BitbucketFilePathClient client;
+    private final String ref;
+
+    /**
+     * Constructor for the root BitbucketSCMFile
+     */
+    public BitbucketSCMFile(BitbucketFilePathClient client, @Nullable String ref) {
+        this.client = client;
+        this.ref = ref;
+        type(DIRECTORY);
+    }
+
+    /**
+     * Constructor for any child file/directory
+     */
+    public BitbucketSCMFile(BitbucketSCMFile parent, String name, Type type) {
+        super(parent, name);
+        client = parent.client;
+        ref = parent.ref;
+        type(type);
+    }
+
+    @Override
+    protected SCMFile newChild(String name, boolean assumeIsDirectory) {
+        return new BitbucketSCMFile(this, name, assumeIsDirectory ? DIRECTORY : REGULAR_FILE);
+    }
+
+    @Override
+    public Iterable<SCMFile> children() throws IOException, InterruptedException {
+        if (isDirectory()) {
+            return client.getDirectoryContent(this);
+        }
+        throw new IOException("Cannot get content- only valid with DIRECTORY type files");
+    }
+
+    public String getFilePath() {
+        String path = getName();
+        SCMFile nextParent = parent();
+        while (nextParent != null && !StringUtils.isEmpty(nextParent.getName())) {
+            path = nextParent.getName() + '/' + path;
```

Comment:
This feels a bit messy but I didn't want to pollute this class with the URL builder. Can be moved to the `FilePathClient` if that's preferred.

Reply:
The file path URL is used by the client as part of the REST URL, so it's fairly critical this method is robust.
Would passing through a URLBuilder, or performing this operation in the file client be preferable?